draft-ietf-quic-transport.md

```diff
@@ -1618,69 +1618,84 @@ of three ways:
 * stateless reset ({{stateless-reset}})
 
 
-### Draining Period {#draining}
+### Draining and Closing Periods {#draining}
 
-After a connection is closed for any reason, an endpoint might receive packets
-from its peer.  These packets might have been sent prior to receiving any close
-signal, or they might be retransmissions of packets for which acknowledgments
-were lost.
+After a connection is closed, an endpoint might receive packets from its peer.
+These packets might have been sent prior to receiving any close signal, or they
+might be retransmissions of packets for which acknowledgments were lost.
 
 The draining period exists to ensure that delayed or reordered packets are
 properly discarded.  This state SHOULD persist for three times the current
-Retransmission Timeout (RTO) interval as defined in {{QUIC-RECOVERY}}.  During
-this period no new application data can be sent on the connection.
+Retransmission Timeout (RTO) interval as defined in {{QUIC-RECOVERY}}.  No
+packets can be sent during this time.
 
-Different treatment is given to packets that are received while a connection is
-in the draining period depending on how the connection was closed.  An endpoint
-that is in a draining period MUST NOT send packets unless they contain a
-CONNECTION_CLOSE or APPLICATION_CLOSE frame.
+Disposing of connection state prior to the end of the draining period could
+result in delayed or reordered packets to be handled poorly.  Endpoints that
+have some alternative means to ensure that late-arriving packets on the
+connection do not create QUIC state, such as by closing the UDP socket, MAY use
+an abbreviated draining period which can allow for faster resource recovery.
+Servers that retain an open socket for accepting new connections SHOULD NOT exit
+the draining period early.
 
```

Comment:
We may want to be a bit more specific about what connection state can be disposed of when entering the closing period. Implementations can probably just drop all information about unacknowledged packets, received packets or streams, but they need to be able to receive packets and decrypt them. If they don't do that, they won't be able to parse the incoming packet and notice arrival of a closing frame.

As for the draining period, state management becomes simpler with the new definition. The implementation just needs to remember the old connection ID, and guarantee that it won't send any packet for that connection ID until the end of the draining -- including not sending any stateless reset packet.

Reply:
I've taken a stab at this.  connection ID is pretty much it.  I will use MAY for the packet protection keys.  I'll observe here (but not in the text) that a key update will be difficult to manage, as much as a path migration is difficult.  Let's not over-optimize for those cases.